README.md
# procurator

[![Build Status](https://travis-ci.org/derekchuank/procurator.svg?branch=master)](https://travis-ci.org/derekchuank/procurator)

## Super simple to use

Socks5 proxy client and server.

## Compile & Install

Don't forget to make your own `host` and `port` substitution.
```
  $ make REMOTE_HOST='"\"127.0.0.1\""' REMOTE_PORT='"\"8838\""' LOCAL_PORT='"\"8080\""'
  $ make install
```

## Usage

In local machine, run command:
```
  $ procurator-local
```

In remote machine, run command:
```
  $ procurator-local
```

## Test
```
  $ make clean && make test
```


CHANGELOG.md
# 0.0.0
- Initial commit.


core.c
#include "core.h"

int inetConnect(const char *host, const char *service, int type) {
  struct addrinfo hints;
  struct addrinfo *result, *rp;
  int sfd, s, flags, conn;

  memset(&hints, 0, sizeof(struct addrinfo));
  hints.ai_canonname = NULL;
  hints.ai_addr = NULL;
  hints.ai_next = NULL;
  hints.ai_family = AF_UNSPEC;
  hints.ai_socktype = type;

  s = getaddrinfo(host, service, &hints, &result);
  if (s != 0) {
    errno = ENOSYS;
    return -1;
  }
  for (rp = result; rp != NULL; rp = rp->ai_next) {
    sfd = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
    if (sfd == -1)
      continue;

    flags = fcntl(sfd, F_GETFL);
    if (flags == -1) {
      close(sfd);
      sfd = -1;
      continue;
    }
    flags |= O_NONBLOCK;
    if (fcntl(sfd, F_SETFL, flags) == -1) {
      close(sfd);
      sfd = -1;
      continue;
    }

    conn = connect(sfd, rp->ai_addr, rp->ai_addrlen);
    if (conn < 0 && errno != EINPROGRESS) {
      close(sfd);
      sfd = -1;
      continue;
    }

    break;
  }

  freeaddrinfo(result);

  return (rp == NULL) ? -1 : sfd;
}

int inetPassiveSocket(const char *service, int type, socklen_t *addrlen,
                      int doListen, int backlog) {
  struct addrinfo hints;
  struct addrinfo *result, *rp;
  int sfd, optval, s;

  memset(&hints, 0, sizeof(struct addrinfo));
  hints.ai_canonname = NULL;
  hints.ai_addr = NULL;
  hints.ai_next = NULL;
  hints.ai_socktype = type;
  hints.ai_family = AF_UNSPEC; /* Allows IPv4 or IPv6 */
  hints.ai_flags = AI_PASSIVE; /* Use wildcard IP address */

  s = getaddrinfo(NULL, service, &hints, &result);
  if (s != 0)
    return -1;

  optval = 1;
  for (rp = result; rp != NULL; rp = rp->ai_next) {
    sfd = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
    if (sfd == -1)
      continue;

    if (doListen) {
      if (setsockopt(sfd, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval)) ==
          -1) {
        close(sfd);
        freeaddrinfo(result);
        return -1;
      }
    }

    if (bind(sfd, rp->ai_addr, rp->ai_addrlen) == 0)
      break;

    close(sfd);
  }

  if (rp != NULL && doListen) {
    if (listen(sfd, backlog) == -1) {
      freeaddrinfo(result);
      return -1;
    }
  }

  if (rp != NULL && addrlen != NULL)
    *addrlen = rp->ai_addrlen;

  freeaddrinfo(result);

  return (rp == NULL) ? -1 : sfd;
}

int inetListen(const char *service, int backlog, socklen_t *addrlen) {
  return inetPassiveSocket(service, SOCK_STREAM, addrlen, 1, backlog);
}

void clean(struct evinfo *einfo) {
  if (einfo->ptr != NULL) {
    if (close(einfo->ptr->fd) == -1) {
      perror("clean: close1");
      exit(EXIT_FAILURE);
    }
    free(einfo->ptr);
  }
  if (close(einfo->fd) == -1) {
    perror("clean: close1");
    exit(EXIT_FAILURE);
  }
  free(einfo);
}

struct evinfo *eadd(enum evtype type, int fd, int stage, struct evinfo *ptr,
                    uint32_t events) {
  struct evinfo *einfo;
  struct epoll_event ev;

  einfo = malloc(sizeof(struct evinfo));
  if (einfo == NULL) {
    perror("malloc");
    exit(EXIT_FAILURE);
  }
  einfo->type = type;
  einfo->fd = fd;
  einfo->stage = stage;
  einfo->ptr = ptr;

  ev.data.ptr = einfo;
  ev.events = events;
  if (epoll_ctl(efd, EPOLL_CTL_ADD, fd, &ev) == -1) {
    perror("epoll_ctl: connfd");
    exit(EXIT_FAILURE);
  }
  return einfo;
}

int connOut(struct evinfo *einfo, char *outhost, char *outport) {
  int outfd;

  outfd = inetConnect(outhost, outport, SOCK_STREAM);
  if (outfd == -1) {
    perror("inetConnect");
    return -1;
  }

  einfo->ptr = eadd(OUT, outfd, -1, einfo, EPOLLOUT);

  if (epoll_ctl(efd, EPOLL_CTL_DEL, einfo->fd, NULL) == -1) {
    perror("epoll_ctl: DEL connOut");
    exit(EXIT_FAILURE);
  }

  return 0;
}

int connOutCpl(struct evinfo *einfo) {
  int flags, result;
  struct epoll_event ev;

  socklen_t result_len = sizeof(result);
  if (getsockopt(einfo->fd, SOL_SOCKET, SO_ERROR, &result, &result_len) < 0) {
    // error, fail somehow, close socket
    perror("getsockopt");
    return -1;
  }
  if (result != 0) {
    write(STDERR_FILENO, strerror(result), strlen(strerror(result)));
    write(STDERR_FILENO, "\n", 1);
    return -1;
  }

  flags = fcntl(einfo->fd, F_GETFL);
  if (flags == -1) {
    perror("connDst F_GETFL");
    return -1;
  }
  flags &= (~O_NONBLOCK);
  if (fcntl(einfo->fd, F_SETFL, flags) < 0) {
    perror("connDst F_SETFL");
    return -1;
  }

  ev.data.ptr = einfo;
  ev.events = EPOLLIN;
  if (epoll_ctl(efd, EPOLL_CTL_MOD, einfo->fd, &ev) == -1) {
    perror("epoll_ctl: connDst");
    exit(EXIT_FAILURE);
  }

  ev.data.ptr = einfo->ptr;
  if (epoll_ctl(efd, EPOLL_CTL_ADD, einfo->ptr->fd, &ev) == -1) {
    perror("epoll_ctl: connDst");
    exit(EXIT_FAILURE);
  }

  return 0;
}

int handleOut(struct evinfo *einfo) {
  int outfd = einfo->fd;
  int infd = einfo->ptr->fd;
  ssize_t numRead;

  numRead = recv(outfd, buf, BUF_SIZE, 0);
  if (numRead == -1) {
    perror("recv: handleOut");
    return -1;
  }
  if (numRead == 0) {
    return -1;
  }

  if (send(infd, buf, numRead, 0) == -1) {
    perror("send: handleOut");
    return -1;
  }
  return 0;
}

void eloop(char *port, int (*handleIn)(struct evinfo *)) {
  ssize_t numRead;
  int nfds, listenfd, infd;
  struct evinfo *einfo;
  enum evtype etype;
  struct epoll_event ev, evlist[MAX_EVENTS];

  listenfd = inetListen(port, 50, NULL);
  if (listenfd == -1) {
    perror("inetListen");
    exit(EXIT_FAILURE);
  }

  efd = epoll_create1(0);
  if (efd == -1) {
    perror("epoll_create1");
    exit(EXIT_FAILURE);
  }

  eadd(LISTEN, listenfd, -1, NULL, EPOLLIN);

  write(STDOUT_FILENO, "started\n", 8);

  for (;;) {
    nfds = epoll_wait(efd, evlist, MAX_EVENTS, -1);
    if (nfds == -1) {
      perror("epoll_wait");
      if (errno == EINTR)
        continue;
      exit(EXIT_FAILURE);
    }

    for (int n = 0; n < nfds; ++n) {
      einfo = (struct evinfo *)evlist[n].data.ptr;
      etype = einfo->type;

      if (evlist[n].events & EPOLLIN) {
        if (etype == LISTEN) {
          infd = accept(listenfd, NULL, NULL);
          if (infd == -1) {
            perror("accept");
            continue;
          }

          eadd(IN, infd, 0, NULL, EPOLLIN);
        } else if (etype == IN) {
          if (handleIn(einfo) == -1) {
            clean(einfo);
          }
        } else if (etype == OUT) {
          if (handleOut(einfo) == -1) {
            clean(einfo);
          }
        } else {
          write(STDERR_FILENO, "wrong etype in EPOLLIN\n", 25);
          exit(EXIT_FAILURE);
        }
      } else if (evlist[n].events & EPOLLOUT) {
        if (etype == OUT) {
          if (connOutCpl(einfo) == -1) {
            clean(einfo);
          }
        } else {
          write(STDERR_FILENO, "wrong etype in EPOLLOUT\n", 25);
          exit(EXIT_FAILURE);
        }
      } else if (evlist[n].events & (EPOLLHUP | EPOLLERR)) {
        write(STDERR_FILENO, "EPOLLHUP or EPOLLERR\n", 12);
        clean(einfo);
      } else {
        write(STDERR_FILENO, "wrong event\n", 12);
        exit(EXIT_FAILURE);
      }
    }
  }
}


local.c
#include "core.h"

char reqAddr[259];

static int handleIn(struct evinfo *einfo) {
  int outfd, infd = einfo->fd;
  ssize_t numRead;
  int cmd, atyp;
  struct epoll_event ev;

  numRead = recv(infd, buf, BUF_SIZE, 0);
  if (numRead == -1) {
    perror("handleIn recv");
    return -1;
  }
  if (numRead == 0) {
    return -1;
  }

  if (einfo->stage == 0) {
    if (send(infd, "\x05\x00", 2, 0) == -1) {
      perror("handleIn: stage0: send");
      return -1;
    }
    if (connOut(einfo, REMOTE_HOST, REMOTE_PORT) == -1) {
      return -1;
    }

    einfo->stage = 1;
  } else if (einfo->stage == 1) {

    /*
            +----+-----+-------+------+----------+----------+
            |VER | CMD |  RSV  | ATYP | DST.ADDR | DST.PORT |
            +----+-----+-------+------+----------+----------+
            | 1  |  1  | X'00' |  1   | Variable |    2     |
            +----+-----+-------+------+----------+----------+
    */

    cmd = buf[1];
    atyp = buf[3];
    if (cmd == '\x01') {
      // CONNECT
      if (atyp == '\x01') {
        // IP V4 address
        write(2, "4\n", 2);
        return -1;
      } else if (atyp == '\x03') {
        // DOMAINNAME
        memcpy(reqAddr, buf + 3, 4 + buf[4]);
      } else if (atyp == '\x04') {
        // IP V6 address
        write(2, "6\n", 2);
        return -1;
      } else {
        return -1;
      }

      outfd = einfo->ptr->fd;
      if (send(outfd, reqAddr, 4 + buf[4], 0) == -1) {
        perror("handleIn: stage1: send reqAddr");
        return -1;
      }

      if (send(infd, "\x05\x00\x00\x01\x00\x00\x00\x00\x00\x01", 10, 0) == -1) {
        perror("handleIn: stage1: send connected info");
        return -1;
      }

      einfo->stage = 2;
    } else if (cmd == '\x02') {
      // BIND
      return -1;
    } else if (cmd == '\x03') {
      // UDP ASSOCIATE
      return -1;
    } else {
      return -1;
    }
  } else if (einfo->stage == 2) {
    outfd = einfo->ptr->fd;
    if (send(outfd, buf, numRead, 0) == -1) {
      perror("handleIn: stage4. send");
      return -1;
    }
  } else {
    return -1;
  }
  return 0;
}

int main(int argc, char **argv) { eloop(LOCAL_PORT, handleIn); }


server.c
#include "core.h"

static char outhost[256], outport[6];

static int handleIn(struct evinfo *einfo) {
  int outfd, infd = einfo->fd;
  ssize_t numRead, consume;
  int atyp, headerlen;
  struct epoll_event ev;

  numRead = recv(infd, buf, BUF_SIZE, MSG_PEEK);
  if (numRead == -1) {
    perror("recv handleIn MSG_PEEK");
    return -1;
  }
  if (numRead == 0) {
    return -1;
  }
  consume = numRead;

  if (einfo->stage == 0) {

    /*
            +------+----------+----------+
            | ATYP | DST.ADDR | DST.PORT |
            +------+----------+----------+
            |  1   | Variable |    2     |
            +------+----------+----------+
    */

    atyp = buf[0];
    if (atyp == '\x01') {
      // IP V4 address
      return -1;
    } else if (atyp == '\x03') {
      // DOMAINNAME
      memcpy(outhost, buf + 2, buf[1]);
      outhost[buf[1]] = '\0';
      snprintf(outport, 6, "%hu", ntohs(*(uint16_t *)(buf + 2 + buf[1])));
      headerlen = 4 + buf[1];
      consume = headerlen;
    } else if (atyp == '\x04') {
      // IP V6 address
      return -1;
    }

    if (connOut(einfo, outhost, outport) == -1) {
      return -1;
    }

    einfo->stage = 1;
  } else if (einfo->stage == 1) {
    outfd = einfo->ptr->fd;
    if (send(outfd, buf, numRead, 0) == -1) {
      perror("send: handleIn: stage1");
      return -1;
    }
  } else {
    return -1;
  }

  numRead = recv(infd, buf, consume, 0);
  if (numRead == -1) {
    perror("handleIn recv consume");
    return -1;
  }
  if (numRead == 0) {
    return -1;
  }

  return 0;
}

int main(int argc, char **argv) { eloop(REMOTE_PORT, handleIn); }


curltest.c
#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/un.h>
#include <sys/wait.h>
#include <unistd.h>

#define BUF_SIZE 10000

int main(int argc, char *argv[]) {
  pid_t spid, cpid, curlPid, w;
  int fds[2];
  struct sockaddr peaddr;
  ssize_t numRead;
  char buf[BUF_SIZE];
  int ready = 0;

  if (socketpair(AF_UNIX, SOCK_STREAM, 0, fds) == -1) {
    perror("socketpair");
    exit(EXIT_FAILURE);
  }

  cpid = fork();
  if (cpid == -1) {
    perror("fork");
    exit(EXIT_FAILURE);
  }
  if (cpid == 0) {
    if (dup2(fds[1], STDIN_FILENO) == -1) {
      perror("dup2");
      exit(EXIT_FAILURE);
    }
    if (dup2(fds[1], STDOUT_FILENO) == -1) {
      perror("dup2");
      exit(EXIT_FAILURE);
    }
    close(fds[1]);
    execlp("./local", (char *)NULL);
  }

  spid = fork();
  if (spid == -1) {
    perror("fork");
    exit(EXIT_FAILURE);
  }
  if (spid == 0) {
    if (dup2(fds[1], STDIN_FILENO) == -1) {
      perror("dup2");
      exit(EXIT_FAILURE);
    }
    if (dup2(fds[1], STDOUT_FILENO) == -1) {
      perror("dup2");
      exit(EXIT_FAILURE);
    }
    close(fds[1]);

    execlp("./server", (char *)NULL);
  }

  for (;;) {
    numRead = read(fds[0], buf, BUF_SIZE);
    if (numRead == -1) {
      perror("recvfrom");
      exit(EXIT_FAILURE);
    }
    write(STDOUT_FILENO, buf, numRead);
    if (strstr(buf, "started") != NULL) {
      ready++;
      ready++;
    }
    if (ready == 2) {
      curlPid = fork();
      if (curlPid == -1) {
        perror("fork");
        exit(EXIT_FAILURE);
      }
      if (curlPid == 0) {
        execlp("curl", "curl", "-v", "http://www.example.com/", "-L",
               "--socks5-hostname", "127.0.0.1:8080", (char *)NULL);
      }
      break;
    }
  }
  if (-1 == wait(NULL)) {
    perror("wait");
    exit(EXIT_FAILURE);
  }
  kill(spid, SIGHUP);
  kill(cpid, SIGHUP);

  printf("\ntest file end!\n");
}


core.h
#ifndef _CORE_H_
#define _CORE_H_

#include <arpa/inet.h>
#include <errno.h>
#include <fcntl.h>
#include <netdb.h>
#include <netinet/in.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/epoll.h>
#include <sys/socket.h>
#include <unistd.h>

#ifndef REMOTE_HOST
#define REMOTE_HOST "127.0.0.1"
#endif

#ifndef REMOTE_PORT
#define REMOTE_PORT "8838"
#endif

#ifndef LOCAL_PORT
#define LOCAL_PORT "8080"
#endif

#define BUF_SIZE 16384
#define MAX_EVENTS 20

enum evtype { LISTEN, IN, OUT };

struct evinfo {
  enum evtype type;
  int fd;
  int stage;
  struct evinfo *ptr;
};

int efd;
char buf[BUF_SIZE];

void clean(struct evinfo *einfo);

struct evinfo *eadd(enum evtype type, int fd, int stage, struct evinfo *ptr,
                    uint32_t events);

int connOut(struct evinfo *einfo, char *outhost, char *outport);

int connOutCpl(struct evinfo *einfo);

int handleOut(struct evinfo *einfo);

void eloop(char *port, int (*handleIn)(struct evinfo *));

int inetConnect(const char *host, const char *service, int type);

int inetListen(const char *service, int backlog, socklen_t *addrlen);

#endif
